### ISSUE TITLE
[Feature] superimpose and digitizer emulation [sf#16]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#40]
**Reported by manuelbi on 2002-06-03 08:46 UTC**
NMS 8280/Sony G900P superimpose emulation via
video-capture-card...

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2003-02-01 12:17 UTC**
Logged In: YES 
user_id=78178

A first idea: implement code that can use the (scaled)
output of another window as input video source.

**Comment by openMSX-import** [comment]
**Updated by manuelbi on 2003-02-01 12:17 UTC**
- **priority**: 5 --> 4

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2005-03-07 17:46 UTC**
Logged In: YES 
user_id=78178

We could also include the Sony HBI-V1 digitizer.

Some information about that is here:
https://www.msx.org/forum/development/msx-development/hbi-v1-programing-information-needed

**Comment by openMSX-import** [comment]
**Updated by manuelbi on 2005-03-07 17:46 UTC**
- **summary**: superimpose emulation with video capture --> superimpose and digitzer emulation

**Comment by openMSX-import** [comment]
**Commented by asiga on 2012-07-09 11:40 UTC**
Please consider implementing this old ticket. I used a NMS8280 for years, and I'd love to be able to have it emulated on openMSX. 

I just saw that the NMS8280 video hardware claims to be emulated by blueMSX, and I believe blueMSX is open source, so it would be possible to take a look at how they did it. 

Btw, if you implement this, don't try to "connect" it to any host video capture device, like blueMSX seems to do (they seem to require that you use a TV grabbing card on the host). Don't do that... it's system-specific, and not too useful. Instead, connect it to generic video files, such as MPEG, motion JPEG, or whatever (don't worry about supporting too many codecs either... just support one video format, and we users will convert to it if necessary).

By using video files instead of system-specific video capture hardware, the possibilities of this feature would be INMENSE :-)

Please, please, please, implement this!!!! :-)))

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2012-07-10 08:15 UTC**
Yeah, my suggestion would also be just that: support superimposing using the background of an OGG/Theora video. We already support playing those videos for the Laserdisc emulation. It would be good enough (at least for a first implementation) to have that as a background video to superimpose on.

asiga: what would you do with this feature? What kind of possibilities are you thinking of? (Such examples may motivate someone to implement it...)

**Comment by openMSX-import** [comment]
**Commented by asiga on 2012-07-10 09:51 UTC**
Yes, the possibilities I'm thinking are turning this NMS8280 feature more comfortable than it was: I remember my old-days digitizing sessions a bit painful because of the cables you had to plug and moreover you had to take care the analog signal had good quality. Some times  VHS tapes created a bad-quality signal which generated distorted digitizing, and it was a pain to tune everything in situations like that. 

I'm imagining being able to feed a perfect signal from a video file into the Videographics software, and being able to use MSX BASIC for creating tools that use the digitizer for doing stuff on such video signal.

Also, I'm imagining some sort of video preprocessor between the video file and the digital input into the NMS8280: First the video will need to be scaled (of course), but it would be possible to add controls for tuning the gamma or brightness (I mean outside of the NMS8280 emulation, like if you were controlling the signal before it arrives to the machine). Also, even the digitizer quality could be enhanced (for example, it would be possible to implement floyd-steinberg dithering from the original 24bit to the 8bit of screen 8)... well, I don't mean all that stuff should be in the first implementation, I'm just seeing the potential.

As a conclusion, I believe this would make MSX2 digitizing apps useful again, like some sort of revival :-)

Asiga

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2012-07-10 09:55 UTC**
Hi,

Thanks for that.

Well, as a reaction... I don't want to temper your enthusiasm, but what is the use of digitizing an already digital signal in an emulator? If you want to convert digital images to screen 8, you can already do that with existing image converters in any quality you like (with dithering, filtering, gamma enhance, etc.) Why would one use VideoGraphics for that via an emulator?

**Comment by openMSX-import** [comment]
**Commented by asiga on 2012-07-10 11:07 UTC**
Well, I'm supposing there's interest in using MSX2 software for doing stuff. For example, I know I could use the latest Photoshop version for drawing sprites for a MSX2 game, but, if there're MSX2 tools that also let me do the work comfortably, why don't use them?

Otherwise it would be like saying that the MSX is for games only. Well, actually, even if it would be for the sake of gaming/demo-ing, the digitizer effects can be cool for writing demos anyway.

Regarding the digital signal being already digital, yes, it's true, but even if you used a video capture card in your host machine, the signal would be already digital... so... by this reasoning, it wouldn't make any sense to digitize from an emulator unless you had some way of accessing analog data from the emulator, and this would be quite shocking... analog data on a digital computer :-)

In conclusion, having the digitizer (+superimpose) features emulated, would allow being able to use the NMS8280 features today, and I believe they would prove not only cool, but also useful (useful if you plan to use MSX2 software, of course).

asiga

**Comment by openMSX-import** [comment]
**Commented by asiga on 2012-07-10 13:37 UTC**
Btw, I've just talked about the digitize feature until now, but I guess the superimpose might convince you more :-)  Imagine that you can feed a video signal into the emulator, use some MSX2 tool for generating superimposed stuff on top of it in real time, and record the result into an output video file... this would be a real use of the emulator taking advantage of these features, wouldn't it ;-)

Btw, the superimpose feature worked over analog video without digitizing it. This makes me believe the best approach for implementing this would be as follows:

1- The video input (a "video file" for example, but could be any other video input if you prefer) is already digital, but it's converted to either PAL or NTSC resolution (in true color, 24 bits per pixel).

2- This 24bpp PAL/NTSC digital signal is what the NMS8280 gets in its input.

3- If superimpose (or video-mix)  is enabled, VDP graphics are superimposed on top of the (unmodified 24bpp) PAL/NTSC image (NOTE: the NMS8280 had a physical slider for controlling the video-mix ratio, as well as the graphics brightness if the mode is superimpose instead of video-mix).

4- If digitizing is enabled, the PAL/NTSC 24bpp signal is "digitized" into 8bpp SCREEN 8 graphics.... I believe you will like this, as this way you're not just feeding a video file, but a PAL/NTSC signal (well, it's already digital, but it's as close as you can get to a real PAL/NTSC digitizer emulator).

Btw, the NMS8280 also had an slider for the brightness of the input signal, which affected both the digitizer and the superimpose features.

Cheers

asiga

**Comment by openMSX-import** [comment]
**Commented by asiga on 2013-03-03 23:07 UTC**
I came here checking if there were any news on NMS8280 emulation. Regarding the question about why digitizing an already digital signal, I see Open MSX already is doing this with sound with the TurboR emulation, because you support ".wav" files as input for audio sampling. I believe it's interesting to have a similar behaviour for emulating the NMS8280 video capabilities.

I've programming knowledge, but I looked at the code for emulating the MSX2 VDP, and I didn't understand the flow clearly (yes, I saw the code for the VDP registers that would control digitizing and superimposing, but I'm not sure how the VDP emulator works). If at least you could leave a couple of empty functions, one for grabbing a frame in MSX2 current screen resolution/pallete, and another for passing a background for superimposing, I'd be happy to code such functions. But I cannot do anything outside such two functions, because reading the code I've no idea on where to write a grabbed frame, or where to put a background for superimposing (btw, the superimpose background should be in the actual resolution/colors of the OpenGL framebuffer, because superimpose is supposed to work with the original video input).

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2013-03-04 20:16 UTC**
There is already superimpose code for the laserdisc emulation (MSX1 graphics are superimposed on top of video from the laserdisc). Check that code to get an idea... Things start in PioneerLDControl.hh and in VDP.cc you'll find a method setExternalVideoSource. You can also check where the LD player frames are coming from and stuff. Just browse the code... or join us on IRC if you need help.
